src/ResolvedPeers.h

```diff
@@ -110,8 +114,53 @@ class ResolvedPeers: public RefCountable
     size_type availablePaths = 0;
 };
 
+/// An invasive reference-counting Comm::Connection pointer that also keeps an
+/// (optional) ResolvedPeers position for the ResolvedPeers::returnPath() usage.
+/// Reference counting mechanism is compatible with Comm::ConnectionPointer.
+class PeerConnectionPointer
+{
+public:
+    using size_type = ResolvedPeers::size_type;
+
+    PeerConnectionPointer() = default;
+    PeerConnectionPointer(nullptr_t): PeerConnectionPointer() {} ///< implicit nullptr conversion
```

Comment:
IMO this would be better written as an assignment operator so the RHS object does not need to go through allocate+deallocate cycles of a nil object just to perform "= nullptr;" assignments.

Reply:
If we add the assignment operator, we would need to add also copy/move constructor, destructor and move operator= , according to the [rule](https://en.wikipedia.org/wiki/Rule_of_three_(C%2B%2B_programming)) of 5. Do you think it is worth burdening this simple class with that?

Reply:
That would be true for adding a default/copy/move assignment operator. They are all defined implicity anyway, otherwise the nullptr_t constructor would not work.

 But I mean an explicit nullptr_t receiving one, like so:
```suggestion
    PeerConnectionPointer &operator =(nullptr_t) { clear(); }
```

I assume that builds okay. If not then the ctor is fine. This is just a minor optimization to avoid some alloc+copy cycles.

Reply:
> IMO this would be better written as an assignment operator so the RHS object does not need to go through allocate+deallocate cycles of a nil object just to perform "= nullptr;" assignments.

Modern optimizing compilers do not actually create a temporary object in this context. It's all inlined. I double checked that expectation by looking at the generated assembly [code](https://godbolt.org/z/mxcBqm).

With some compilers, adding an explicit nullptr assignment operator will generate slightly different code so I cannot claim that adding that operator makes no difference at all in this case. The optimizations that a compiler can do will differ a bit. I speculate that the resulting performance difference is very small, but this analysis exceeds my assembly interpretation capabilities.

Adding a nullptr assignment operator will duplicate some code -- there is currently no duplication in this area because all "clearing" code reuses default member initialization instead of spelling out how those members should be "cleared".

If Amos wants an explicit nullptr assignment operator added after reading the above, I do not object to that addition despite the (minor) code duplication that addition will cause.

Reply:
Please note that the nullptr assignment operator should _not_ replace the existing from-nullptr constructor. The existing conversion constructor should remain because we want to continue to support both operations (i.e. construction and assignment), at least. Adding an explicit nullptr assignment operator is nothing but an optimization.

Reply:
Hmm. okay since the alloc is being removed these days that clears my concern.